docs/2024/puzzles/day14.md

```diff
@@ -1,11 +1,312 @@
+
+## Part 1
+
+Part 1 shouldn't be too bad. Let's get started with our `Robot` class (and a `Vec2i` class):
+
+```scala
+case class Vec2i(x: Int, y: Int)
+
+case class Robot(pos: Vec2i, velocity: Vec2i)
+```
+
+
+Now we can parse our input: 
+
+```scala
+def parse(str: String): List[Robot] =
+  str.linesIterator.map:
+    case s"p=$px,$py v=$vx,$vy" =>
+      Robot(Vec2i(px.toInt, py.toInt), Vec2i(vx.toInt, vy.toInt))
+  .toList
+```
+
+Let's define our grid size in a `val`, as it can depend on if we are doing a test input or not:
+
+```scala
+val size = Vec2i(101, 103)
+```
+
+The problem text states that when a robot goes off the edge it comes back on the other side, which sounds a lot like modulo.
+Unfortunately, modulo is incorrect in our case for negative numbers. Let's define a remainder instead:
+
+```scala
+extension (self: Int)
+  infix def rem(that: Int): Int =
+    val m = math.abs(self) % that
+    if self < 0 then
+      that - m
+    else
+      m
+```
+
+Now we can add a function to `Robot` that advances the state by `n`:
+
+```scala
+case class Robot(pos: Vec2i, velocity: Vec2i):
+  def stepN(n: Int = 1): Robot =
+    copy(pos = pos.copy(x = (pos.x + n * velocity.x) rem size.x, y = (pos.y + n * velocity.y) rem size.y))
+```
+
+Now for the full `List[Robot]`, we can add `stepN` to that too, and also define our safety function:
+
+```scala
+extension (robots: List[Robot])
+  def stepN(n: Int = 1): List[Robot] = robots.map(_.stepN(n))
+
+  def safety: Int =
+    val middleX = size.x / 2
+    val middleY = size.y / 2
+
+    robots.groupBy: robot =>
+      (robot.pos.x.compareTo(middleX), robot.pos.y.compareTo(middleY)) match
+        case (0, _) | (_, 0) => -1
+        case ( 1, -1) => 0
+        case (-1, -1) => 1
+        case (-1,  1) => 2
+        case ( 1,  1) => 3
+    .removed(-1).values.map(_.length).product
+```
+
+Let's explain this a little. There are 4 quadrants, and as specified there are also lines that aren't in any quadrant.
+We can use `groupBy` to group the robots into quadrants.
+
+First, we get the midpoints by dividing the size by 2. We then compare the robot to the midpoint, returning `-1` if it's on the line
+(either comparison is equal), and otherwise sort the remaining 4 results into quadrants. We then remove the robots on the line,
+get the length of each of the lists, and multiply them together.
+
+With this `part1` is easy to implement:
+
+```scala
+def part1(input: String): Int = parse(input).stepN(100).safety
+```
+
+## Part 2
+
+Part 2 wants us to find an image which is really hard. Thankfully, there is one thing I know about Christmas trees: They have
+a lot of lines in a row and are an organized shape.
+
+The christmas trees I know are really tall, but only in a few columns. So let's test the vertical columns for a few lines that are really long.
+They also have a lot of shorter horizontal lines, so let's also check the rows for a lot of shorter lines.
```

Comment:
Another assumption here is that the tree itself is solid; made up of many robots adjacent to each other. It's not like, an outline of a tree shape.

Reply:
I don't know how it would be possible to detect that if that was the case. Probably something along the lines of a flood fill but that would cost too much

Reply:
Agreed, yeah. I meant moreso it's worth being explicit and listing it as an assumption here.

Reply:
Added section beforehand specifying assumption being made